src/devices_models/devices/common/device_constraint_inputs.jl

```diff
@@ -0,0 +1,129 @@
+struct ModelRangeConstraintInputs
+    constraint_name::String
+    variable_name::String
+    bin_variable_name::Union{Nothing, String}
+    limits_func::Function
+    constraint_func::Function
+end
+
+function ModelRangeConstraintInputs(;
+    constraint_name,
+    variable_name,
+    bin_variable_name = nothing,
+    limits_func,
+    constraint_func,
+)
+    return ModelRangeConstraintInputs(
+        constraint_name,
+        variable_name,
+        bin_variable_name,
+        limits_func,
+        constraint_func,
+    )
+end
+
+struct ModelTimeSeriesConstraintInputs
+    constraint_name::String
+    variable_name::String
+    bin_variable_name::Union{Nothing, String}
+    parameter_name::Union{Nothing, String}
+    forecast_label::Union{Nothing, String}
+    multiplier_func::Union{Nothing, Function}
+    constraint_func::Function
+end
+
+function ModelTimeSeriesConstraintInputs(;
+    constraint_name,
+    variable_name,
+    bin_variable_name = nothing,
+    parameter_name,
+    forecast_label,
+    multiplier_func,
+    constraint_func,
+)
+    return ModelTimeSeriesConstraintInputs(
+        constraint_name,
+        variable_name,
+        bin_variable_name,
+        parameter_name,
+        forecast_label,
+        multiplier_func,
+        constraint_func,
+    )
+end
+
+struct DeviceConstraintInputs
+    range_constraint_inputs::Vector{ModelRangeConstraintInputs}
+    timeseries_range_constraint_inputs::Vector{ModelTimeSeriesConstraintInputs}
+    custom_psi_container_func::Union{Nothing, Function}
+end
+
+function DeviceConstraintInputs(;
+    range_constraint_inputs = Vector{ModelRangeConstraintInputs}(),
+    timeseries_range_constraint_inputs = Vector{ModelTimeSeriesConstraintInputs}(),
+    custom_psi_container_func = nothing,
+)
+    return DeviceConstraintInputs(
+        range_constraint_inputs,
+        timeseries_range_constraint_inputs,
+        custom_psi_container_func,
+    )
+end
+
+function device_constraints!(
+    psi_container::PSIContainer,
+    devices::IS.FlattenIteratorWrapper{T},
+    model::DeviceModel{T, U},
+    feedforward::Union{Nothing, AbstractAffectFeedForward},
+    inputs::DeviceConstraintInputs,
+) where {T <: PSY.Device, U <: AbstractDeviceFormulation}
+    model_range_constraints = inputs.range_constraint_inputs
+    model_timeseries_range_constraints = inputs.timeseries_range_constraint_inputs
+    custom_psi_container_func = inputs.custom_psi_container_func
+
+    # TODO: Could be faster if we iterate over the devices first.
```

Comment:
I believe that this will make a jump model that has constraints in order that you create them. If I'm right about that, there could be value in the looping order being as is. I.e. all ramp limit constraints are adjacent in the jump model. I think it would be much harder to read a jump model if the constraints were grouped by device rather than by constraint type... but it's really just a jump model readability thing.

Reply:
it isn't just a readability thing. We can't push into the DenseAxisArrays to add new constraints. It will also make the results extraction far more difficult due to the sparse nature of the alternative.